### PR TITLE
Add auto static/dynamic

### LINKED ISSUE #7355: [RFC] Rendering in Next.js (SSR, Pre-rendering, CSR, SPA)
Next.js currently has 2 modes of rendering:

- Dynamic rendering means render on demand when a request comes in.
- Pre-rendering means rendering to html at build time using `next export`

These modes work fine if your application only has pages of a certain type. For example [zeit.co/docs](http://zeit.co/docs) is completely static. However more often than not your application is not binary, and requirements for static rendering change over time.

For example, [zeit.co](http://zeit.co) has a blog with static content, marketing pages, a dashboard and more. In the current model, zeit.co would be deployed with the serverless target and every page becomes a serverless function, including the blog and marketing pages that can actually be generated at build time.

Furthermore we've seen a common pattern where you'd want the dashboard to be an appshell type application that only shows the header and a loading spinner in the pre-rendered response, then after hydrating the page the data is fetched and rendered to make the application feel faster. This is generally referred to as a client-side rendered application.

Since Next.js is in control of the compilation pipeline we can decide at build time if a page will always get the same result by inferring if the page has data requirements. If it doesn't have data requirements we'll automatically render the page as a static HTML file.

When this proposal is implemented you'll be able to choose between pre-rendering and dynamic rendering on the page level. 

## Goals

- Provide fast by default experience
- Optimize pages that we know are always going to render the same result
- Allow usage of dynamic, static or client-side only on a per-page level
- Support client-side only applications that lazy-load data
- Powerful use-case for dashboards
- Great for marketing pages that lazy-load login / authentication state

## API

This proposal doesn't need any API changes, it's mostly related to changing the semantics and internals of Next.js to export HTML in certain cases during `next build`.

Initially, we will cover the case where `getInitialProps` is not defined. Meaning that if a page doesn't have `getInitialProps` it is automatically exported as HTML.

```js
// pages/about.js
export default () => <p>Hello world</p>
```

This is always going to render the same, so during build, we export it to an HTML file.

We might also need to detect router usage, but we won't cover that in the initial implementation as it's not as common. In this case you probably don't want to do dynamic rendering.

```js
import { useRouter } from 'next/router'

export default () => {
  const router = useRouter()
  const { query } = router
  
  return <p>Page: {query.page}</p>
}
```

On the Now side of things `@now/next` will be updated to upload the `.html` files generated by the build.

### PR DESCRIPTION
**Pull request by ijjk**
This adds the start of automatically turning pages without dynamic content into static files. 

Closes: #7355

### CODE REVIEW

File: packages/next-server/server/next-server.ts [hunk 1]
```diff
@@ -319,7 +319,15 @@ export default class Server {
     query: ParsedUrlQuery = {},
     opts: any,
   ) {
+    // try serving a static AMP version first
+    if (query.amp) {
+      try {
+        const result = await loadComponents(this.distDir, this.buildId, (pathname === '/' ? '/index' : pathname) + '.amp')
+        if (typeof result.Component === 'string') return result.Component
+      } catch (_) {}
```
Comment (timneutkens) [inline packages/next-server/server/next-server.ts]: This should handle the errors coming from the execution.

Comment (ijjk) [reply packages/next-server/server/next-server.ts]: Updated

File: packages/next/build/index.ts [hunk 2]
```diff
@@ -271,7 +271,7 @@ export default async function build(dir: string, conf = null): Promise<void> {
       actualPage + '.js'
     )
 
-    if (!page.match(/(_app|_error|_document)/)) {
+    if (!page.match(/(_app|_error|_document|^\/api)/)) {
```
Comment (timneutkens) [inline packages/next/build/index.ts]: This is missing `/^` so it'll match `_app` anywhere in the string.

Comment (ijjk) [reply packages/next/build/index.ts]: Updated

**Review by Timer — APPROVED** [review]
I'm comfortable merging this because it's behind a feature flag. 🙏

**Pull request merged**